rosidl_typesupport_c/src/type_support_dispatch.hpp

```diff
@@ -34,85 +33,103 @@ namespace rosidl_typesupport_c
-        } catch (const std::exception & e) {
-          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-            "Failed to compute library name for '%s' due to %s",
-            library_basename, e.what());
-          return nullptr;
-        }
-
-        try {
-          lib = new rcpputils::SharedLibrary(library_name);
-        } catch (const std::runtime_error & e) {
-          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-            "Could not load library %s: %s", library_name.c_str(), e.what());
-          return nullptr;
-        } catch (const std::bad_alloc & e) {
-          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-            "Could not load library %s: %s", library_name.c_str(), e.what());
-          return nullptr;
-        }
-        map->data[i] = lib;
+  if (handle->typesupport_identifier != rosidl_typesupport_c__typesupport_identifier) {
+    return nullptr;
+  }
+
+  const type_support_map_t * map =
+    static_cast<const type_support_map_t *>(handle->data);
+  for (size_t i = 0; i < map->size; ++i) {
+    if (strcmp(map->typesupport_identifier[i], identifier) != 0) {
+      continue;
+    }
+    rcpputils::SharedLibrary * lib = nullptr;
+
+    if (!map->data[i]) {
+      char library_basename[1024];
+      int ret = rcutils_snprintf(
+        library_basename, 1023, "%s__%s",
+        map->package_name, identifier);
+      if (ret < 0) {
+        RCUTILS_SET_ERROR_MSG("Failed to format library name");
+        return nullptr;
       }
-      auto clib = static_cast<const rcpputils::SharedLibrary *>(map->data[i]);
-      lib = const_cast<rcpputils::SharedLibrary *>(clib);
 
-      void * sym = nullptr;
+      std::string library_name;
+      try {
+        library_name = rcpputils::get_platform_library_name(library_basename);
+      } catch (const std::runtime_error & e) {
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Failed to compute library name for '%s' due to %s",
+          library_basename, e.what());
+        return nullptr;
+      }
 
       try {
-        if (!lib->has_symbol(map->symbol_name[i])) {
-          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-            "Failed to find symbol '%s' in library", map->symbol_name[i]);
-          return nullptr;
-        }
-        sym = lib->get_symbol(map->symbol_name[i]);
-      } catch (const std::exception & e) {
+        lib = new rcpputils::SharedLibrary(library_name);
+      } catch (const std::runtime_error & e) {
         RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-          "Failed to get symbol '%s' in library: %s",
-          map->symbol_name[i], e.what());
+          "Could not load library %s: %s", library_name.c_str(), e.what());
+        return nullptr;
+      } catch (const std::bad_alloc & e) {
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Could not load library %s: %s", library_name.c_str(), e.what());
         return nullptr;
       }
+      map->data[i] = lib;
```

Comment:
Any explanation of how this code works that you're aware of? It's a little strange that it modifies a non-const `void **` member of a `const type_support_t *`.

Reply:
Honestly, I just moved it around.  @hidmic was involved in the latest rewrite of it, I believe.  Maybe he can shed some light.